From c9c74cc2545affd65da4d5095d900f185d05f33c Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 26 Jul 2013 22:19:29 -0400
Subject: [PATCH] Allow force spawns

Pass spawn reason force to not allow plugins to block
---
 src/main/java/net/minecraft/server/Entity.java | 8 ++------
 src/main/java/net/minecraft/server/World.java  | 4 ++--
 2 files changed, 4 insertions(+), 8 deletions(-)

diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 6fba047..5bcafaf 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -14,8 +14,8 @@ import org.bukkit.block.BlockFace;
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.entity.Painting;
 import org.bukkit.entity.Vehicle;
+import org.bukkit.event.entity.*;
 import org.spigotmc.CustomTimingsHandler; // Spigot
-import org.bukkit.event.entity.EntityCombustByEntityEvent;
 import org.bukkit.event.painting.PaintingBreakByEntityEvent;
 import org.bukkit.event.vehicle.VehicleBlockCollisionEvent;
 import org.bukkit.event.vehicle.VehicleEnterEvent;
@@ -23,10 +23,6 @@ import org.bukkit.event.vehicle.VehicleExitEvent;
 import org.bukkit.craftbukkit.CraftWorld;
 import org.bukkit.craftbukkit.entity.CraftEntity;
 import org.bukkit.craftbukkit.entity.CraftPlayer;
-import org.bukkit.event.entity.EntityCombustEvent;
-import org.bukkit.event.entity.EntityDamageByBlockEvent;
-import org.bukkit.event.entity.EntityDamageEvent;
-import org.bukkit.event.entity.EntityPortalEvent;
 import org.bukkit.plugin.PluginManager;
 // CraftBukkit end
 
@@ -1833,7 +1829,7 @@ public abstract class Entity {
                     entity.setPositionRotation((double) chunkcoordinates.x, (double) chunkcoordinates.y, (double) chunkcoordinates.z, entity.yaw, entity.pitch);
                 }
                 // CraftBukkit end */
-                worldserver1.addEntity(entity);
+                worldserver1.addEntity(entity, CreatureSpawnEvent.SpawnReason.FORCE); // EMC
                 // CraftBukkit start - Forward the CraftEntity to the new entity
                 this.getBukkitEntity().setHandle(entity);
                 entity.bukkitEntity = this.getBukkitEntity();
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 8193fb9..a6ca053 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -977,13 +977,13 @@ public abstract class World implements IBlockAccess {
             }
         } // Spigot end
 
-        if (event != null && (event.isCancelled() || entity.dead)) {
+        if (spawnReason != SpawnReason.FORCE && event != null && (event.isCancelled() || entity.dead)) { // EMC
             entity.dead = true;
             return false;
         }
         // CraftBukkit end
 
-        if (!flag && !this.isChunkLoaded(i, j)) {
+        if (spawnReason != SpawnReason.FORCE && !flag && !this.isChunkLoaded(i, j)) { // EMC
             entity.dead = true; // CraftBukkit
             return false;
         } else {
-- 
1.8.3.4

